### PR TITLE
wgengine: inject packetbuffers rather than bytes

### LINKED ISSUE #2741: wgengine/netstack: remove alloc-per-packet
The netstack mode data path includes this full copy+alloc per packet:

```go
func (ns *Impl) injectOutbound() {
        for {
                packetInfo, ok := ns.linkEP.ReadContext(context.Background())
                if !ok {
                        ns.logf("[v2] ReadContext-for-write = ok=false")
                        continue
                }
                pkt := packetInfo.Pkt
                hdrNetwork := pkt.NetworkHeader()
                hdrTransport := pkt.TransportHeader()

                full := make([]byte, 0, pkt.Size())
                full = append(full, hdrNetwork.View()...)
                full = append(full, hdrTransport.View()...)
                full = append(full, pkt.Data().AsRange().AsView()...)
                if debugNetstack {
                        ns.logf("[v2] packet Write out: % x", full)
                }
                if err := ns.tundev.InjectOutbound(full); err != nil {
                        log.Printf("netstack inject outbound: %v", err)
                        return
                }

        }
}
```

That was lazy get-it-working code and known to be inefficient at the time, but we haven't gotten back to improving it yet.

/cc @stapelberg @crawshaw

### PR DESCRIPTION
**Pull request by raggi**
( child change of #4222 )

Plumb the outbound injection path to allow passing netstack
PacketBuffers down to the tun Read, where they are decref'd to enable
buffer re-use. This removes one packet alloc & copy, and reduces GC
pressure by pooling outbound injected packets.

Fixes #2741

The following before and after heap samples are taken from a fresh instance of tailscaled running in userspace mode, transporting a 383mb file over http between lan connected nodes:

Before:
```
Showing nodes accounting for 1.92GB, 96.19% of 2GB total
Dropped 234 nodes (cum <= 0.01GB)
Showing top 10 nodes out of 58
      flat  flat%   sum%        cum   cum%
    0.54GB 27.10% 27.10%     0.54GB 27.10%  golang.zx2c4.com/wireguard/device.(*Device).PopulatePools.func1
    0.39GB 19.45% 46.55%     0.41GB 20.50%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).readFromPayloader
    0.38GB 18.87% 65.42%     0.38GB 18.87%  tailscale.com/wgengine/netstack.(*Impl).injectOutbound
    0.28GB 14.04% 79.46%     0.28GB 14.04%  gvisor.dev/gvisor/pkg/tcpip/stack.glob..func1
    0.16GB  7.84% 87.29%     0.16GB  7.84%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*segment).clone (inline)
    0.08GB  4.16% 91.46%     0.09GB  4.46%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.newIncomingSegment
    0.04GB  2.24% 93.69%     0.04GB  2.24%  sync.(*Pool).pinSlow
    0.02GB  1.13% 94.82%     0.31GB 15.43%  gvisor.dev/gvisor/pkg/tcpip/stack.NewPacketBuffer
    0.01GB  0.71% 95.53%     0.01GB  0.71%  gvisor.dev/gvisor/pkg/tcpip/stack.(*IPTables).CheckOutput
    0.01GB  0.66% 96.19%     0.01GB  0.66%  sync.(*poolChain).pushHead
 ```
 
 After:
 ```
 Showing nodes accounting for 1060.55MB, 94.50% of 1122.22MB total
Dropped 179 nodes (cum <= 5.61MB)
Showing top 10 nodes out of 71
      flat  flat%   sum%        cum   cum%
  400.05MB 35.65% 35.65%   409.05MB 36.45%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*endpoint).readFromPayloader
  296.27MB 26.40% 62.05%   296.27MB 26.40%  golang.zx2c4.com/wireguard/device.(*Device).PopulatePools.func1
  171.58MB 15.29% 77.34%   171.58MB 15.29%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.(*segment).clone (inline)
   97.55MB  8.69% 86.03%   102.55MB  9.14%  gvisor.dev/gvisor/pkg/tcpip/transport/tcp.newIncomingSegment
   26.50MB  2.36% 88.39%    34.51MB  3.08%  gvisor.dev/gvisor/pkg/tcpip/stack.NewPacketBuffer
   24.60MB  2.19% 90.58%    24.60MB  2.19%  sync.(*Pool).pinSlow
      14MB  1.25% 91.83%       14MB  1.25%  sync.(*poolChain).pushHead
      12MB  1.07% 92.90%       12MB  1.07%  gvisor.dev/gvisor/pkg/tcpip/stack.(*IPTables).CheckOutput
    9.50MB  0.85% 93.75%     9.50MB  0.85%  gvisor.dev/gvisor/pkg/tcpip/stack.(*IPTables).CheckPostrouting
    8.50MB  0.76% 94.50%   129.06MB 11.50%  tailscale.com/wgengine/netstack.(*Impl).injectInbound
 ```

### CODE REVIEW

File: net/tstun/wrap.go [hunk 1]
```diff
@@ -494,15 +499,22 @@ func (t *Wrapper) Read(buf []byte, offset int) (int, error) {
...
-		// We are done with t.buffer. Let poll re-use it.
-		t.sendBufferConsumed()
+
+	var n int
+	if res.packet != nil {
+		n = copy(buf[offset:], res.packet.NetworkHeader().View())
+		n += copy(buf[offset+n:], res.packet.TransportHeader().View())
+		n += copy(buf[offset+n:], res.packet.Data().AsRange().AsView())
+
+		defer res.packet.DecRef()
```
Comment (maisem) [inline net/tstun/wrap.go:509]: why defer?

**Comment by bradfitz** [comment]
misc notes, mostly on commit message style:

* drop the /gvisor from commit message title. That's not a package name.
* Put "Updates #nnn" or "Fixes #nnnn" in commit messages on their own lines, depending on whether it's a full fix or partial work towards the fix (Updates). In PR messages it's optional. It's the commit messages we care about.
* we tend to use the word "bump" instead of "roll". like "go.mod: bump netstack".
*

File: net/tstun/wrap.go [hunk 2]
```diff
@@ -158,8 +159,12 @@ type Wrapper struct {
 // The byte slice is not interpreted in the usual way for a Read method.
 // See the comment in the middle of Wrap.Read.
 type tunReadResult struct {
-	data []byte
-	err  error
+	// Only one of data or packet should be set, if set packet takes precedence.
```
Comment (bradfitz) [inline net/tstun/wrap.go:162]: or err?

File: net/tstun/wrap.go [hunk 3]
```diff
@@ -158,8 +159,12 @@ type Wrapper struct {
...
+	packet *stack.PacketBuffer
+	data   []byte
+	err    error
+
+	injected bool
```
Comment (bradfitz) [inline net/tstun/wrap.go:167]: and what does this mean?

**Review by bradfitz — COMMENTED** [review]
Nice :)

**Comment by bradfitz** [comment]
Can we do a simple go.mod netstack bump alone without any other changes first? Then all the code changes in one?

File: net/tstun/wrap.go (hunk 2 above)
Comment (raggi) [reply net/tstun/wrap.go]: 👍 clarified the comment.

File: net/tstun/wrap.go (hunk 3 above)
Comment (raggi) [reply net/tstun/wrap.go:170]: 👍 added a comment

File: net/tstun/wrap.go (hunk 1 above)
Comment (raggi) [reply net/tstun/wrap.go]: Turns out this came from a rerere, which I've turned off with glee - quite right there was no need for it to be a defer.

File: net/tstun/wrap.go [hunk 4]
```diff
@@ -154,12 +155,19 @@ type Wrapper struct {
 	disableTSMPRejected bool
 }
 
-// tunReadResult is the result of a TUN read: Some data and an error.
-// The byte slice is not interpreted in the usual way for a Read method.
+// tunReadResult is the result of a TUN read, or an injected result pretending to be a tun read.
```
Comment (bradfitz) [inline net/tstun/wrap.go:158]: TUN vs tun case within a few words of each other. It's not an acronym I believe but it seems we say TUN in English text.

Comment (bradfitz) [reply net/tstun/wrap.go:158]: (as a variable, just follow normal Go https://github.com/golang/go/wiki/CodeReviewComments#initialisms rules)

File: net/tstun/wrap.go [hunk 5]
```diff
@@ -741,10 +756,26 @@ func (t *Wrapper) InjectOutbound(packet []byte) error {
...
+	size := packet.Size()
+	if size > MaxPacketSize {
+		return errPacketTooBig
+	}
+	if size == 0 {
+		return nil
+	}
+	t.sendOutbound(tunReadResult{packet: packet, injected: true})
+	return nil
+
```
Comment (bradfitz) [inline net/tstun/wrap.go:776]: extra blank line

File: net/tstun/wrap.go [hunk 6]
```diff
@@ -741,10 +756,26 @@ func (t *Wrapper) InjectOutbound(packet []byte) error {
...
 	return nil
 }
 
+// InjectOutboundPacketBuffer logically behaves as InjectOutbound. It takes ownership of one
+// reference count on the packet, and the packet may be mutated. The packet refcount will be
+// decremented after the injected buffer has been read.
+func (t *Wrapper) InjectOutboundPacketBuffer(packet *stack.PacketBuffer) error {
+	size := packet.Size()
+	if size > MaxPacketSize {
+		return errPacketTooBig
```
Comment (bradfitz) [inline net/tstun/wrap.go:769]: missing DecRef?

File: net/tstun/wrap.go [hunk 7]
```diff
@@ -741,10 +756,26 @@ func (t *Wrapper) InjectOutbound(packet []byte) error {
...
+	}
+	if size == 0 {
+		return nil
```
Comment (bradfitz) [inline net/tstun/wrap.go:772]: missing DecRef?

File: wgengine/netstack/netstack.go [hunk 8]
```diff
@@ -382,17 +382,12 @@ func (ns *Impl) injectOutbound() {
...
-		full = append(full, hdrNetwork.View()...)
-		full = append(full, hdrTransport.View()...)
-		full = append(full, pkt.Data().AsRange().AsView()...)
 		if debugPackets {
-			ns.logf("[v2] packet Write out: % x", full)
+			ns.logf("[v2] packet Write out: % x", stack.PayloadSince(pkt.NetworkHeader()))
 		}
-		if err := ns.tundev.InjectOutbound(full); err != nil {
+
+		if err := ns.tundev.InjectOutboundPacketBuffer(pkt); err != nil {
```
Comment (bradfitz) [inline wgengine/netstack/netstack.go:390]: Maybe a tiny comment here explaining the refcount ownership transfer?

// pkt has non-zero refcount; InjectOutboundPacketBuffer will decrement...

**Review by bradfitz — APPROVED** [review]

File: net/tstun/wrap.go (hunk 4 above)
Comment (danderson) [reply net/tstun/wrap.go:158]: "tun" is just short for "tunnel", not an acronym. I think we inherited the "TUN" misspelling from WireGuard code, but the correct go format is "Tun".

File: net/tstun/wrap.go (hunk 6 above)
Comment (raggi) [reply net/tstun/wrap.go:769]: 👍

File: net/tstun/wrap.go (hunk 7 above)
Comment (raggi) [reply net/tstun/wrap.go:772]: 👍

File: net/tstun/wrap.go (hunk 4 above)
Comment (raggi) [reply net/tstun/wrap.go:158]: I made it slightly more self-consistent, with the type name always lowercased, and docstrings always uppercased.

**Comment by bradfitz** [comment]
Rebase down to one commit?

**Comment by raggi** [comment]
> Rebase down to one commit?

yeah, it is, the parent commit is the go.mod one, pushed the rebase for clarity :)

**Review by bradfitz — APPROVED** [review]
LGTM! And ran it here locally and verified the basics all work.

**Pull request merged**